File: backend/internal/core/entities/veterinary.go
package entities

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// VeterinaryVisitType represents the type of veterinary visit
type VeterinaryVisitType string

const (
	VisitTypeCheckup     VeterinaryVisitType = "checkup"
	VisitTypeVaccination VeterinaryVisitType = "vaccination"
	VisitTypeTreatment   VeterinaryVisitType = "treatment"
	VisitTypeSurgery     VeterinaryVisitType = "surgery"
	VisitTypeEmergency   VeterinaryVisitType = "emergency"
)

// VeterinaryVisit represents a veterinary visit
type VeterinaryVisit struct {
	ID            primitive.ObjectID  `bson:"_id,omitempty" json:"id"`
	AnimalID      string              `bson:"animal_id" json:"animal_id"`
	AnimalName    string              `bson:"animal_name" json:"animal_name"`
	Type          VeterinaryVisitType `bson:"type" json:"type"`
	VisitDate     time.Time           `bson:"visit_date" json:"visit_date"`
	VeterinarianID string             `bson:"veterinarian_id,omitempty" json:"veterinarian_id,omitempty"`
	Veterinarian  string              `bson:"veterinarian" json:"veterinarian"`
	Clinic        string              `bson:"clinic,omitempty" json:"clinic,omitempty"`
	Diagnosis     string              `bson:"diagnosis,omitempty" json:"diagnosis,omitempty"`
	Treatment     string              `bson:"treatment,omitempty" json:"treatment,omitempty"`
	Medications   []Medication        `bson:"medications,omitempty" json:"medications,omitempty"`
	NextVisitDate *time.Time          `bson:"next_visit_date,omitempty" json:"next_visit_date,omitempty"`
	Cost          float64             `bson:"cost,omitempty" json:"cost,omitempty"`
	Notes         string              `bson:"notes,omitempty" json:"notes,omitempty"`
	CreatedAt     time.Time           `bson:"created_at" json:"created_at"`
	UpdatedAt     time.Time           `bson:"updated_at" json:"updated_at"`
	RecordedBy    string              `bson:"recorded_by" json:"recorded_by"`
}

// Vaccination represents a vaccination record
type Vaccination struct {
	ID              primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	AnimalID        string             `bson:"animal_id" json:"animal_id"`
	AnimalName      string             `bson:"animal_name" json:"animal_name"`
	VaccineName     string             `bson:"vaccine_name" json:"vaccine_name"`
	VaccineType     string             `bson:"vaccine_type" json:"vaccine_type"`
	DateAdministered time.Time         `bson:"date_administered" json:"date_administered"`
	NextDueDate     *time.Time         `bson:"next_due_date,omitempty" json:"next_due_date,omitempty"`
	BatchNumber     string             `bson:"batch_number,omitempty" json:"batch_number,omitempty"`
	Veterinarian    string             `bson:"veterinarian" json:"veterinarian"`
	Clinic          string             `bson:"clinic,omitempty" json:"clinic,omitempty"`
	Reactions       string             `bson:"reactions,omitempty" json:"reactions,omitempty"`
	Notes           string             `bson:"notes,omitempty" json:"notes,omitempty"`
	CreatedAt       time.Time          `bson:"created_at" json:"created_at"`
	RecordedBy      string             `bson:"recorded_by" json:"recorded_by"`
}

// Medication represents a medication
type Medication struct {
	Name        string     `bson:"name" json:"name"`
	Dosage      string     `bson:"dosage" json:"dosage"`
	Frequency   string     `bson:"frequency" json:"frequency"`
	StartDate   time.Time  `bson:"start_date" json:"start_date"`
	EndDate     *time.Time `bson:"end_date,omitempty" json:"end_date,omitempty"`
	Instructions string    `bson:"instructions,omitempty" json:"instructions,omitempty"`
}

// HealthCondition represents an ongoing health condition
type HealthCondition struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	AnimalID    string             `bson:"animal_id" json:"animal_id"`
	Condition   string             `bson:"condition" json:"condition"`
	DiagnosedDate time.Time        `bson:"diagnosed_date" json:"diagnosed_date"`
	Status      string             `bson:"status" json:"status"` // active, resolved, chronic
	Treatment   string             `bson:"treatment,omitempty" json:"treatment,omitempty"`
	Notes       string             `bson:"notes,omitempty" json:"notes,omitempty"`
	CreatedAt   time.Time          `bson:"created_at" json:"created_at"`
	UpdatedAt   time.Time          `bson:"updated_at" json:"updated_at"`
}

// VeterinaryVisitCreateRequest represents veterinary visit creation
type VeterinaryVisitCreateRequest struct {
	AnimalID       string              `json:"animal_id" validate:"required"`
	Type           VeterinaryVisitType `json:"type" validate:"required,oneof=checkup vaccination treatment surgery emergency"`
	VisitDate      time.Time           `json:"visit_date" validate:"required"`
	VeterinarianID string              `json:"veterinarian_id,omitempty"`
	Veterinarian   string              `json:"veterinarian" validate:"required"`
	Clinic         string              `json:"clinic,omitempty"`
	Diagnosis      string              `json:"diagnosis,omitempty"`
	Treatment      string              `json:"treatment,omitempty"`
	Medications    []Medication        `json:"medications,omitempty"`
	NextVisitDate  *time.Time          `json:"next_visit_date,omitempty"`
	Cost           float64             `json:"cost,omitempty" validate:"omitempty,gte=0"`
	Notes          string              `json:"notes,omitempty"`
}

// VaccinationCreateRequest represents vaccination creation
type VaccinationCreateRequest struct {
	AnimalID         string     `json:"animal_id" validate:"required"`
	VaccineName      string     `json:"vaccine_name" validate:"required"`
	VaccineType      string     `json:"vaccine_type" validate:"required"`
	DateAdministered time.Time  `json:"date_administered" validate:"required"`
	NextDueDate      *time.Time `json:"next_due_date,omitempty"`
	BatchNumber      string     `json:"batch_number,omitempty"`
	Veterinarian     string     `json:"veterinarian" validate:"required"`
	Clinic           string     `json:"clinic,omitempty"`
	Reactions        string     `json:"reactions,omitempty"`
	Notes            string     `json:"notes,omitempty"`
}

// NewVeterinaryVisit creates a new veterinary visit
func NewVeterinaryVisit(animalID, animalName string, visitType VeterinaryVisitType, visitDate time.Time, veterinarian, recordedBy string) *VeterinaryVisit {
	now := time.Now()
	return &VeterinaryVisit{
		ID:           primitive.NewObjectID(),
		AnimalID:     animalID,
		AnimalName:   animalName,
		Type:         visitType,
		VisitDate:    visitDate,
		Veterinarian: veterinarian,
		CreatedAt:    now,
		UpdatedAt:    now,
		RecordedBy:   recordedBy,
	}
}

// NewVaccination creates a new vaccination record
func NewVaccination(animalID, animalName, vaccineName, vaccineType string, dateAdministered time.Time, veterinarian, recordedBy string) *Vaccination {
	now := time.Now()
	return &Vaccination{
		ID:               primitive.NewObjectID(),
		AnimalID:         animalID,
		AnimalName:       animalName,
		VaccineName:      vaccineName,
		VaccineType:      vaccineType,
		DateAdministered: dateAdministered,
		Veterinarian:     veterinarian,
		CreatedAt:        now,
		RecordedBy:       recordedBy,
	}
}

File: frontend/src/types/veterinary.ts
export interface VeterinaryVisit {
  id: string
  animal_id: string
  animal?: {
    id: string
    name: string
    species: string
  }
  visit_date: string
  visit_type: 'checkup' | 'emergency' | 'surgery' | 'vaccination' | 'follow_up' | 'other'
  veterinarian_name: string
  clinic_name?: string
  reason: string
  diagnosis?: string
  treatment_provided?: string
  medications_prescribed?: string
  follow_up_required: boolean
  follow_up_date?: string
  weight?: number
  temperature?: number
  heart_rate?: number
  notes?: string
  cost?: number
  attachments?: string[]
  created_at?: string
  updated_at?: string
}

export interface Vaccination {
  id: string
  animal_id: string
  animal?: {
    id: string
    name: string
    species: string
  }
  vaccine_name: string
  vaccine_type: string
  manufacturer?: string
  batch_number?: string
  vaccination_date: string
  next_due_date?: string
  veterinarian_name: string
  clinic_name?: string
  site_of_injection?: string
  dosage?: string
  adverse_reactions?: string
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface Medication {
  id: string
  animal_id: string
  animal?: {
    id: string
    name: string
    species: string
  }
  medication_name: string
  medication_type: 'antibiotic' | 'pain_relief' | 'anti_inflammatory' | 'antiparasitic' | 'supplement' | 'other'
  dosage: string
  frequency: string
  route: 'oral' | 'topical' | 'injection' | 'intravenous' | 'other'
  start_date: string
  end_date?: string
  prescribed_by: string
  reason: string
  instructions?: string
  side_effects?: string
  status: 'active' | 'completed' | 'discontinued'
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface TreatmentPlan {
  id: string
  animal_id: string
  animal?: {
    id: string
    name: string
    species: string
  }
  plan_name: string
  condition: string
  start_date: string
  end_date?: string
  veterinarian_name: string
  goals: string
  medications?: string[]
  procedures?: string[]
  dietary_requirements?: string
  exercise_requirements?: string
  monitoring_schedule?: string
  progress_notes?: Array<{
    date: string
    note: string
    recorded_by: string
  }>
  status: 'active' | 'completed' | 'cancelled'
  created_at?: string
  updated_at?: string
}

export interface MedicalCondition {
  id: string
  animal_id: string
  animal?: {
    id: string
    name: string
    species: string
  }
  condition_name: string
  diagnosis_date: string
  diagnosed_by: string
  severity: 'mild' | 'moderate' | 'severe' | 'critical'
  status: 'active' | 'managed' | 'resolved' | 'chronic'
  symptoms?: string
  treatment?: string
  prognosis?: string
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface VeterinaryStatistics {
  total_visits: number
  upcoming_vaccinations: number
  active_medications: number
  active_treatment_plans: number
  active_conditions: number
  visits_this_month: number
  vaccinations_due_soon: number
}
